circuit/src/gadgets/common.rs

```diff
@@ -0,0 +1,751 @@
+		bit: AssignedCell<F, F>,
+		// Assigns a cell for the x.
+		x: AssignedCell<F, F>,
+		// Assigns a cell for the y.
+		y: AssignedCell<F, F>,
+		config: CommonConfig,
+		mut layouter: impl Layouter<F>,
+	) -> Result<AssignedCell<F, F>, Error> {
+		// Checking bit is boolean or not.
+		let assigned_bool = Self::is_bool(bit, config, layouter.namespace(|| "is_boolean"))?;
+
+		layouter.assign_region(
+			|| "select",
+			|mut region: Region<'_, F>| {
+				config.selectors[5].enable(&mut region, 0)?;
+
+				let assigned_bit =
+					assigned_bool.copy_advice(|| "bit", &mut region, config.advice[0], 0)?;
+				let assigned_x = x.copy_advice(|| "x", &mut region, config.advice[1], 0)?;
+				let assigned_y = y.copy_advice(|| "y", &mut region, config.advice[2], 0)?;
+
+				// Conditional control checks the bit. Is it zero or not?
+				// If yes returns the y value, else x.
+				let res = assigned_bit.value().and_then(|bit_f| {
+					if bool::from(bit_f.is_zero()) {
+						assigned_y.value().cloned()
+					} else {
+						assigned_x.value().cloned()
+					}
+				});
+
+				let assigned_res = region.assign_advice(|| "res", config.advice[1], 1, || res)?;
+
+				Ok(assigned_res)
+			},
+		)
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+	use crate::utils::{generate_params, prove_and_verify};
+	use halo2wrong::{
+		curves::bn256::{Bn256, Fr},
+		halo2::{
+			circuit::{SimpleFloorPlanner, Value},
+			dev::MockProver,
+			plonk::{Circuit, Instance},
+		},
+	};
+
+	#[derive(Clone)]
+	struct TestConfig {
+		common: CommonConfig,
+		temp: Column<Advice>,
+		pub_ins: Column<Instance>,
+	}
+
+	#[derive(Clone)]
+	struct TestCircuit<F: FieldExt, const N: usize> {
+		inputs: [F; N],
+		c: char,
+	}
+
+	impl<F: FieldExt, const N: usize> TestCircuit<F, N> {
+		fn new(inputs: [F; N], c: char) -> Self {
+			Self { inputs, c }
+		}
+	}
+
+	impl<F: FieldExt, const N: usize> Circuit<F> for TestCircuit<F, N> {
+		type Config = TestConfig;
+		type FloorPlanner = SimpleFloorPlanner;
+
+		fn without_witnesses(&self) -> Self {
+			self.clone()
+		}
+
+		fn configure(meta: &mut ConstraintSystem<F>) -> TestConfig {
+			let common = CommonChip::configure(meta);
+			let temp = meta.advice_column();
+			let pub_ins = meta.instance_column();
+			meta.enable_equality(temp);
+			meta.enable_equality(pub_ins);
+
+			TestConfig { common, pub_ins, temp }
+		}
+
+		fn synthesize(
+			&self, config: TestConfig, mut layouter: impl Layouter<F>,
+		) -> Result<(), Error> {
+			let mut items = Vec::new();
+			for i in 0..N {
+				items.push(layouter.assign_region(
+					|| "temp",
+					|mut region: Region<'_, F>| {
+						let x = region.assign_advice(
+							|| "temp_inputs",
+							config.temp,
+							i,
+							|| Value::known(self.inputs[i]),
+						)?;
+						Ok(x)
+					},
+				)?);
+			}
+
+			match self.c {
+				'a' => {
+					let and = CommonChip::and(
+						items[0].clone(),
+						items[1].clone(),
+						config.common,
+						layouter.namespace(|| "and"),
+					)?;
```

Comment:
Why did you decide to use a char to match, instead of using an `enum`. e.g.:
```rust
enum Gadgets {
  ...
  And(F, F),
  IsBool(F),
  ...
}
```

Reply:
In your example, I need to carry [assign_region](https://github.com/eigen-trust/eigen-trust/blob/common-gadget/circuit/src/gadgets/common.rs#L408) inside each match case to access enums data if I am not missing anything at rust knowledge side.
Maybe  `enum Gadgets{
		And,
		IsBool,
		IsEqual,
		IsZero,
		Mul,
		Select
	}` will be more aesthetic.
In this case input will be something like this: `let test_chip = TestCircuit::new([Fr::from(1), Fr::from(1)], Gadgets::And);`
I used char in a similar manner, now I think the enum variant seems more aesthetic.

Reply:
Ah, right. That solution sounds good